distrobuilder/main.go

```diff
@@ -637,13 +637,14 @@ fix_systemd_override_unit() {
 	mkdir -p "${dropin_dir}"
 	{
 		echo "[Service]";
-		[ "${systemd_version}" -ge 247 ] && echo "ProtectProc=default";
-		[ "${systemd_version}" -ge 232 ] && echo "ProtectControlGroups=no";
-		[ "${systemd_version}" -ge 232 ] && echo "ProtectKernelTunables=no";
 		[ "${systemd_version}" -ge 249 ] && echo "LoadCredential=";
 
-		# Additional settings for privileged containers
-		if grep -q 4294967295 /proc/self/uid_map; then
```

Comment:
Hmm, with this change, you're now going to be adding a whole bunch of config keys like `ProtectHome=no` to unprivileged containers that don't have nesting enabled whereas they didn't need those in the past.

Reply:
I think we still need the check for whether the container is unprivileged to restrict those keys that only make sense for privileged containers.

Reply:
> Hmm, with this change, you're now going to be adding a whole bunch of config keys like `ProtectHome=no` to unprivileged containers that don't have nesting enabled whereas they didn't need those in the past.

The thing is `ProtectHome` is IMHO currently broken in unprivileged containers that don't have nesting enabled. Here's a focal container without nesting:

```
root@focal-ctn:~# systemctl show systemd-resolved --value -p MainPID
106
root@focal-ctn:~# systemctl show systemd-resolved --value -p ProtectHome
yes
root@focal-ctn:~# nsenter -m -t $(systemctl show systemd-resolved --value -p MainPID) mount | grep home
tmpfs on /run/systemd/unit-root/home type tmpfs (rw,nosuid,nodev,size=3279544k,mode=755,uid=1000000,gid=1000000,inode64)
root@focal-ctn:~# nsenter -m -t $(systemctl show systemd-resolved --value -p MainPID) ls -l /home
total 1
drwxr-xr-x 2 ubuntu ubuntu 5 Oct  4 13:52 ubuntu
```

`/home` should be an empty tmpfs but it is not.

After enabling nesting and restarting the ctn:

```
root@focal-ctn:~# systemctl show systemd-resolved --value -p MainPID
106
root@focal-ctn:~# systemctl show systemd-resolved --value -p ProtectHome
yes
root@focal-ctn:~# nsenter -m -t $(systemctl show systemd-resolved --value -p MainPID) mount | grep home
tmpfs on /home type tmpfs (ro,nosuid,nodev,size=3279544k,mode=755,uid=1000000,gid=1000000,inode64)
root@focal-ctn:~# nsenter -m -t $(systemctl show systemd-resolved --value -p MainPID) ls -l /home
total 0
```

This time it behaved OK.

Reply:
Is the same true for all the other features we turn off in privileged containers but currently keep in unprivileged?

This is a pretty big and scary hammer we use already, I don't want it to do anything more than it needs.

Reply:
Everything that relies on a mount namespace is affected, so all the `Protect*` and `Private*` features of systemd.

Reply:
Hmm, I find that hard to believe as mount namespaces are absolutely allowed to be created even when security.nesting isn't set to true.

Basically all that `security.nesting` does in an unprivileged container is allow all mounts. When not set, unprivileged containers get "almost all mounts" thanks to a massive list of allowed mount flag combinations, then privileged containers get very restricted as they don't get that massive list of allowed mount flag combinations.

e.g. of an unprivileged container without nesting enabled having ProtectHome working just fine:
```
lxc launstgraber@dakara:~/data/code/lxc/lxd-pkg-snap (latest-candidate)$ lxc launch images:ubuntu/20.04 u1
Creating u1
Starting u1
stgraber@dakara:~/data/code/lxc/lxd-pkg-snap (latest-candidate)$ lxc exec u1 bash
root@u1:~# ps aux | grep resolv
systemd+     129  0.0  0.0  23968  9292 ?        Ss   20:00   0:00 /lib/systemd/systemd-resolved
root         261  0.0  0.0  10768   528 pts/1    S+   20:00   0:00 grep --color=auto resolv
root@u1:~# nsenter -t 129 -m
# grep home /proc/self/mountinfo
8984 8983 0:169 /systemd/inaccessible/dir /home ro,nosuid,nodev shared:2415 - tmpfs tmpfs rw,size=6512172k,mode=755,uid=1000000,gid=1000000,inode64
# 
```

Reply:
Now with your patch, the above container would get the real /home exposed to that resolved process, that's a security regression.

Reply:
Here on a 20.04 host with 5.11, I cannot reproduce what you shared:

```
sdeziel@sdeziel-lemur:~$ lxc launch ubuntu/20.04 u1
Creating u1
The local image 'ubuntu/20.04' couldn't be found, trying 'ubuntu:20.04' instead.
Starting u1                                 
sdeziel@sdeziel-lemur:~$ lxc shell u1
root@u1:~# ps aux| grep resolved
systemd+     303  0.0  0.0  23968  9480 ?        Ss   20:18   0:00 /lib/systemd/systemd-resolved
root         905  0.0  0.0   8168   536 pts/0    S+   20:19   0:00 grep --color=auto resolved
root@u1:~# nsenter -t 303 -m
root@u1:/# grep home /proc/self/mountinfo
3705 3704 0:117 /systemd/inaccessible/dir /run/systemd/unit-root/home rw,nosuid,nodev - tmpfs tmpfs rw,size=3279544k,mode=755,uid=1000000,gid=1000000,inode64
root@u1:/# ll /home/ubuntu/
total 15
drwxr-xr-x 3 ubuntu ubuntu    6 Oct  4 20:19 ./
drwxr-xr-x 3 root   root      3 Oct  4 20:19 ../
-rw-r--r-- 1 ubuntu ubuntu  220 Feb 25  2020 .bash_logout
-rw-r--r-- 1 ubuntu ubuntu 3771 Feb 25  2020 .bashrc
-rw-r--r-- 1 ubuntu ubuntu  807 Feb 25  2020 .profile
drwx------ 2 ubuntu ubuntu    3 Oct  4 20:19 .ssh/
```

Reply:
```
stgraber@dakara:~$ lxc launch images:ubuntu/20.04/cloud v1 --vm
Creating v1
Starting v1
stgraber@dakara:~$ lxc exec v1 bash
root@v1:~# apt install snapd
Reading package lists... 7%
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  apparmor liblzo2-2 squashfs-tools
Suggested packages:
  apparmor-profiles-extra apparmor-utils zenity | kdialog
The following NEW packages will be installed:
  apparmor liblzo2-2 snapd squashfs-tools
0 upgraded, 4 newly installed, 0 to remove and 0 not upgraded.
Need to get 31.1 MB of archives.
After this operation, 136 MB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 apparmor amd64 2.13.3-7ubuntu5.1 [494 kB]
Get:2 http://archive.ubuntu.com/ubuntu focal/main amd64 liblzo2-2 amd64 2.10-2 [50.8 kB]
Get:3 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 squashfs-tools amd64 1:4.4-1ubuntu0.2 [117 kB]
Get:4 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 snapd amd64 2.51.1+20.04ubuntu2 [30.4 MB]
Fetched 31.1 MB in 2s (13.5 MB/s)
Preconfiguring packages ...
Selecting previously unselected package apparmor.
(Reading database ... 50099 files and directories currently installed.)
Preparing to unpack .../apparmor_2.13.3-7ubuntu5.1_amd64.deb ...
Unpacking apparmor (2.13.3-7ubuntu5.1) ...
Selecting previously unselected package liblzo2-2:amd64.
Preparing to unpack .../liblzo2-2_2.10-2_amd64.deb ...
Unpacking liblzo2-2:amd64 (2.10-2) ...
Selecting previously unselected package squashfs-tools.
Preparing to unpack .../squashfs-tools_1%3a4.4-1ubuntu0.2_amd64.deb ...
Unpacking squashfs-tools (1:4.4-1ubuntu0.2) ...
Selecting previously unselected package snapd.
Preparing to unpack .../snapd_2.51.1+20.04ubuntu2_amd64.deb ...
Unpacking snapd (2.51.1+20.04ubuntu2) ...
Setting up liblzo2-2:amd64 (2.10-2) ...
Setting up apparmor (2.13.3-7ubuntu5.1) ...
Created symlink /etc/systemd/system/sysinit.target.wants/apparmor.service → /lib/systemd/system/apparmor.service.
Reloading AppArmor profiles 
Skipping profile in /etc/apparmor.d/disable: usr.sbin.rsyslogd
Setting up squashfs-tools (1:4.4-1ubuntu0.2) ...
Setting up snapd (2.51.1+20.04ubuntu2) ...
Created symlink /etc/systemd/system/multi-user.target.wants/snapd.apparmor.service → /lib/systemd/system/snapd.apparmor.service.
Created symlink /etc/systemd/system/multi-user.target.wants/snapd.autoimport.service → /lib/systemd/system/snapd.autoimport.service.
Created symlink /etc/systemd/system/multi-user.target.wants/snapd.core-fixup.service → /lib/systemd/system/snapd.core-fixup.service.
Created symlink /etc/systemd/system/multi-user.target.wants/snapd.recovery-chooser-trigger.service → /lib/systemd/system/snapd.recovery-chooser-trigger.service.
Created symlink /etc/systemd/system/multi-user.target.wants/snapd.seeded.service → /lib/systemd/system/snapd.seeded.service.
Created symlink /etc/systemd/system/cloud-final.service.wants/snapd.seeded.service → /lib/systemd/system/snapd.seeded.service.
Created symlink /etc/systemd/system/multi-user.target.wants/snapd.service → /lib/systemd/system/snapd.service.
Created symlink /etc/systemd/system/timers.target.wants/snapd.snap-repair.timer → /lib/systemd/system/snapd.snap-repair.timer.
Created symlink /etc/systemd/system/sockets.target.wants/snapd.socket → /lib/systemd/system/snapd.socket.
Created symlink /etc/systemd/system/final.target.wants/snapd.system-shutdown.service → /lib/systemd/system/snapd.system-shutdown.service.
snapd.failure.service is a disabled or a static unit, not starting it.
snapd.snap-repair.service is a disabled or a static unit, not starting it.
Processing triggers for libc-bin (2.31-0ubuntu9.2) ...
Processing triggers for systemd (245.4-4ubuntu3.13) ...
Processing triggers for dbus (1.12.16-2ubuntu2.1) ...
Processing triggers for mime-support (3.64ubuntu1) ...
root@v1:~# snap install lxd
2021-10-04T20:43:56Z INFO Waiting for automatic snapd restart...
Warning: /snap/bin was not found in your $PATH. If you've not restarted your session since you
         installed snapd, try doing that. Please see https://forum.snapcraft.io/t/9469 for more
         details.

lxd 4.18 from Canonical✓ installed
root@v1:~# exit
stgraber@dakara:~$ lxc exec v1 bash
root@v1:~# lxd init --auto
root@v1:~# lxc launch images:ubuntu/20.04 u1
Creating u1
Starting u1                                 
root@v1:~# lxc exec u1 bash    
root@u1:~# ps aux | grep resolv
systemd+     140  0.0  1.2  23960 12144 ?        Ss   20:44   0:00 /lib/systemd/systemd-resolved
root         257  0.0  0.0  10760   664 pts/0    S+   20:44   0:00 grep --color=auto resolv
root@u1:~# grep home /proc/140/mountinfo
1149 1148 0:70 /systemd/inaccessible/dir /home ro,nosuid,nodev shared:546 - tmpfs tmpfs rw,size=100088k,mode=755,uid=1000000,gid=1000000
root@u1:~# 
```

Reply:
I finally found the smoking gun:

1) I setup a VM like your `v1` but setup lxd to use ZFS as storage backend
2) creating `u1` works and systemd-resolved's mount feature works

```
root@v1:~# lxc start u1
root@v1:~# lxc exec u1 bash
root@u1:~# ps aux| grep resolv
systemd+     119  0.0  0.9  23968  9460 ?        Ss   23:12   0:00 /lib/systemd/systemd-resolved
root         141  0.0  0.0  10768   504 pts/0    S+   23:12   0:00 grep --color=auto resolv
root@u1:~# nsenter -t 119 -m
# grep home /proc/self/mountinfo
1148 1147 0:72 /systemd/inaccessible/dir /home ro,nosuid,nodev shared:600 - tmpfs tmpfs rw,size=99836k,mode=755,uid=1000000,gid=1000000,inode64
```

However, if I set noatime on the storage backend, systemd-resolved sees a writable home:

```
root@v1:~# zfs set atime=off default
root@v1:~# lxc stop u1
root@v1:~# lxc start u1
root@v1:~# lxc exec u1 bash
root@u1:~# ps aux| grep resolv
systemd+     119  0.0  0.9  23968  9408 ?        Ss   23:14   0:00 /lib/systemd/systemd-resolved
root         141  0.0  0.0  10768   508 pts/0    S+   23:14   0:00 grep --color=auto resolv
root@u1:~# nsenter -t 119 -m
# grep home /proc/self/mountinfo
1147 1146 0:72 /systemd/inaccessible/dir /run/systemd/unit-root/home rw,nosuid,nodev - tmpfs tmpfs rw,size=99836k,mode=755,uid=1000000,gid=1000000,inode64
```

Toggling the `atime` confirms it's really the source of the problem.

Note: I've been setting `atime=off` since forever for performance and storage efficiency.

Reply:
As suggested by stgraber on IRC, this seems to be due to the Apparmor profile missing the noatime combination:

```
root@v1:~# journalctl -b-1 -k -o cat --grep apparmor | grep resolved
audit: type=1400 audit(1633389260.926:49): apparmor="DENIED" operation="mount" info="failed flags match" error=-13 profile="lxd-u1_</var/snap/lxd/common/lxd>" name="/run/systemd/unit-root/" pid=6329 comm="(resolved)" flags="ro, remount, noatime, bind"
audit: type=1400 audit(1633389792.465:58): apparmor="DENIED" operation="mount" info="failed flags match" error=-13 profile="lxd-u1_</var/snap/lxd/common/lxd>" name="/run/systemd/unit-root/" pid=7451 comm="(resolved)" flags="ro, remount, noatime, bind"
```

Reply:
Could it be because the noatime variant is not in https://github.com/lxc/lxd/blob/master/lxd/apparmor/instance_lxc.go#L551-L563 ?

Reply:
That's definitely possible. Feel free to send a PR to add it and we'll see if that does it :)